File: brain/data_processing/layout.py
from dataclasses import dataclass
from typing import List

from .base_layout import BaseLayout


###------------------------------------------------------------------------------------------###
#
#   If you plan to add or remove some data in plugin's recorder remember to add them here and always increment
#   SUPPORTED_LAYOUT_VERSION (also, the order of variables MUST MATCH 100%)
#   String variables, like enemy playmakers, MUST be stored inside NamedStatesContainer as vocabulary will be created only
#   for this class.
#
#   Supported Types: int, float, bool_input, NamedStatesContainer and lists containing those - any of those should automatically be converted and
#   standardized. Any variable with unsupported type will just get skipped.
#
#   Difference between continuous types (i.e. int-s and float-s) and the booleans is that the booleans won't be standardized
#   in processing step as they are well-behaved in the first place. It's also important to note that for the
#   behavioral learning step the boolean-type variables of the recording's user inputs get weights assigned.
#   Without it the network will have trouble learning to "use the buttons" so it's very important to implement everything
#   correctly here.
#
###------------------------------------------------------------------------------------------###

class LayoutNotSupported(Exception):
    pass

SUPPORTED_LAYOUT_VERSION: int = 16

@dataclass
class NamedState(BaseLayout):
    Name: str
    StateName: str

@dataclass
class NamedStatesContainer(BaseLayout):
    ParentName: str # for example enemy name
    NamedStates: List[NamedState] # for example list of enemy playmakers

@dataclass
class Hero(BaseLayout):
    PosX: float
    PosY: float
    RelPosX: float
    RelPosY: float
    HP: int
    Silk: int
    IsStunned: bool
    CanJump: bool
    CanDoubleJump: bool
    CanAttack: bool
    CanSprint: bool
    CanBind: bool
    CanCast: bool
    CanNailArt: bool
    CanTryHarpoon: bool
    CanBackDash: bool

    InvDistX: float
    InvDistY: float

@dataclass
class Enemy(BaseLayout):
    PosX: float
    PosY: float
    RelPosX: float
    RelPosY: float
    VelX: float
    VelY: float
    HP: int
    Facing: bool
    PlayMakers: NamedStatesContainer

@dataclass
class UserInputs(BaseLayout):
    Horizontal: float
    Vertical: float

    Jump: bool
    Attack: bool
    Heal: bool
    Skill: bool
    Dash: bool
    Harpoon: bool

@dataclass
class FrameData(BaseLayout):
    Hero: Hero
    Enemies: List[Enemy]

@dataclass
class RecordingFrame(BaseLayout):
    FrameData: FrameData
    Targets: UserInputs
    Reward: int

@dataclass
class LiveInferenceFrame(BaseLayout):
    FrameData: FrameData
    Reward: int

File: plugin/AIPlugin/DataCollection/Layout.cs
using AIPlugin.BossfightSession;
using AIPlugin.Networking;
using AIPlugin.Utilities;
using MessagePack;


namespace AIPlugin //TODO fix all namespaces
{
    [MessagePackObject]
    public class RecordingHeader
    {
        [Key("format_version")] public int FormatVersion = 17;
        [Key("record_frame_delta")] public int RecordFrameDelta = SessionConfig.CaptureFrameDelta;
        [Key("target_name")] public string BossName;
        [Key("enemy_names")] public string[] EnemyNames;
        [Key("player_name")] public string PlayerName;
    }
    [MessagePackObject]
    public class RecordingFooter
    {
        [Key("success")] public bool Success;
        [Key("frame_count")] public int FrameCount;
        [Key("total_reward")] public int TotalReward;
    }

    [MessagePackObject]
    public class FrameData
    {
        [Key(0)] public FrameHeroData Hero;
        [Key(1)] public FrameEnemyData[] Enemies; // The main target (i.e. boss) should always be at the first index
    }

    [MessagePackObject]
    public class RecordingFrame
    {
        [Key(0)] public FrameData Data;
        [Key(1)] public FrameUserInputs UserInputs;
        [Key(2)] public int Reward = 0;

    }
    [MessagePackObject]
    public class InferenceFrame: IPayload
    {
        [Key(0)] public FrameData Data;
        [Key(1)] public int Reward = 0;
        public static InferenceFrame FromRecordingFrameData(RecordingFrame recordingFrame)
        {
            return new InferenceFrame()
            {
                Data = recordingFrame.Data,
                Reward = recordingFrame.Reward,
            };
        }
    }

    [MessagePackObject]
    public struct FrameEnemyData
    {
        [Key(0)] public float PosX;
        [Key(1)] public float PosY;
        [Key(2)] public float RelPosX; // relative to hero
        [Key(3)] public float RelPosY; 
        [Key(4)] public float VelX;
        [Key(5)] public float VelY;
        [Key(6)] public int HP;
        [Key(7)] public bool Facing; //(true)->x_scale >= 0
        [Key(8)] public NamedStatesContainer PlayMakers; //PlayMakerFSM.FsmName and PlayMakerFSM.ActiveStateName
    }
    [MessagePackObject]
    public struct NamedStatesContainer
    {
        [Key(0)] public string ParentName;
        [Key(1)] public NamedState[] NamedStates;
    }

    [MessagePackObject]
    public struct NamedState
    {
        [Key(0)] public string Name; 
        [Key(1)] public string StateName; 
    }

    [MessagePackObject]
    public struct FrameHeroData 
    {
        [Key(0)] public float PosX;
        [Key(1)] public float PosY;
        [Key(2)] public float RelPosX; // relative to targetEnemy
        [Key(3)] public float RelPosY;
        [Key(4)] public int HP;
        [Key(5)] public int Silk;
        [Key(6)] public bool IsStunned;
        [Key(7)] public bool CanJump;
        [Key(8)] public bool CanDoubleJump;
        [Key(9)] public bool CanAttack;
        [Key(10)] public bool CanSprint;
        [Key(11)] public bool CanBind;
        [Key(12)] public bool CanCast;
        [Key(13)] public bool CanNailArt;
        [Key(14)] public bool CanTryHarpoon;
        [Key(15)] public bool CanBackDash;
        [Key(16)] public float InvDistX;
        [Key(17)] public float InvDistY;
    }
    [MessagePackObject]
    public class FrameUserInputs: IResponse
    {
        // Movement:
        [Key(0)] public float Horizontal;
        [Key(1)] public float Vertical;
        [Key(2)] public bool Jump;
        //Actions:
        [Key(3)] public bool Attack;
        [Key(4)] public bool Heal;
        [Key(5)] public bool Skill;
        [Key(6)] public bool Dash;
        [Key(7)] public bool Harpoon;
    }
}